### PR TITLE
emit envelops to metric server

### PR DESCRIPTION
**Pull request by qibobo**

### CODE REVIEW

File: src/autoscaler/metricsgateway/emitter.go [hunk 1]
```diff
@@ -1,36 +1,68 @@
...
+		handshakeTimeout:     handshakeTimeout,
 		envelopChan:          make(chan *loggregator_v2.Envelope, bufferSize),
+		doneChan:             make(chan bool),
+		dialer: websocket.Dialer{
+			TLSClientConfig:  tlsConfig,
+			Proxy:            http.ProxyFromEnvironment,
+			HandshakeTimeout: handshakeTimeout,
+		},
+		eclock:                     eclock,
+		verifyWSConnectionInterval: verifyWSConnectionInterval,
```
Comment (boyang9527) [inline src/autoscaler/metricsgateway/emitter.go]: is it KeepAlive?

**Comment by boyang9527** [comment]
we need to add logic to reconnect if the ws connection fails.

**Comment by qibobo** [comment]
@boyang9527 how to check the ws con fail? existed verifyWSConnection is not enough? Or when there is err in readMessage or writeMessage means the con fail?

**Comment by qibobo** [comment]
@boyang9527  updated

File: src/autoscaler/metricsgateway/cmd/metricsgateway/main.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		nozzles[i] = metricsgateway.NewNozzle(logger, i, shardID, rlpAddr, loggregatorClientTLSConfig, envelopChan, getAppIDsFunc)
+	}
+	return nozzles
+}
+
+func createEmitters(logger lager.Logger, bufferSize int, eclock clock.Clock, keepAliveInterval time.Duration, metricsServerAddrs []string, metricServerClientTLSConfig *tls.Config, handshakeTimeout time.Duration) []metricsgateway.Emitter {
+	emitters := make([]metricsgateway.Emitter, len(metricsServerAddrs))
+	for i := 0; i < len(metricsServerAddrs); i++ {
+		wsHelper := mg_helpers.NewWSHelper(metricsServerAddrs[i]+routes.EnvelopePath, metricServerClientTLSConfig, handshakeTimeout, logger)
+		emitter := metricsgateway.NewEnvelopeEmitter(logger, bufferSize, eclock, keepAliveInterval, wsHelper)
```
Comment (boyang9527) [inline src/autoscaler/metricsgateway/cmd/metricsgateway/main.go]: merge these two lines

File: src/autoscaler/metricsgateway/config/config.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	DefaultEnvelopChanSize                  = 500
+	DefaultEmitterBufferSize                = 500
+)
+
+type AppManagerConfig struct {
+	AppRefreshInterval time.Duration     `yaml:"app_refresh_interval"`
+	PolicyDB           db.DatabaseConfig `yaml:"policy_db"`
+}
+type DispatcherConfig struct {
+	AppRefreshInterval time.Duration `yaml:"app_refresh_interval"`
```
Comment (boyang9527) [inline src/autoscaler/metricsgateway/config/config.go]: why we have this conf for dispatcher?

File: src/autoscaler/metricsgateway/helpers/ws_helper.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	wh.logger.Info("setup-new-ws-connection")
+	URL, err := url.Parse(wh.metricServerURL)
+	if err != nil {
+		return err
+	}
+
+	if URL.Scheme != "wss" && URL.Scheme != "ws" {
+		return fmt.Errorf("Invalid scheme '%s'", URL.Scheme)
+	}
+	retryCount := 1
```
Comment (boyang9527) [inline src/autoscaler/metricsgateway/helpers/ws_helper.go]: what the retryCount is used for? given you don't have max retry

File: src/autoscaler/metricsgateway/helpers/ws_helper.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	CloseConn() error
+	Write(envelope *loggregator_v2.Envelope) error
+	Read() error
+	Ping() error
+}
+
+const (
+	DefaultMaxSetupRetryCount = 10
+	DefaultMaxCloseRetryCount = 10
+	DefaultRetryDelay         = 500 * time.Millisecond
```
Comment (boyang9527) [inline src/autoscaler/metricsgateway/helpers/ws_helper.go]: it is too short

**Comment by boyang9527** [comment]
please address the comments and rebase from gateway branch

**Comment by qibobo** [comment]
@boyang9527  updated.

**Comment by boyang9527** [comment]
I will merge to gateway branch first.. but we need to consolidate  ws_helper.go with web socket test helpers in metricscollector.  Most of the code are duplicated. This can be addressed by another PR

**Pull request merged**